src/procedural-macros.md

```diff
@@ -1,24 +1,546 @@
+
+#[proc_macro_derive(MyTrait)]
+pub fn foo(item: TokenStream) -> TokenStream {
+    println!("{:#?}", item);
+    "fn answer() -> u32 { 2 }".parse().unwrap()
+}
+```
+
+using it liek:
+
+```rust,ignore
+extern crate my_macro;
+
+use my_macro::MyTrait;
+
+#[derive(MyTrait)]
+struct Foo;
+
+fn main() {
+    drop(Foo);
+    println!("the answer was: {}", answer());
+}
+```
+
+and compiling it:
+
+```sh
+$ cargo run
+   Compiling my-macro v0.1.0 (file://.../foo/my-macro)
+   Compiling foo v0.1.0 (file://.../foo)
+TokenStream [
+    Ident { ident: "struct", span: #0 bytes(67..73) },
+    Ident { ident: "Foo", span: #0 bytes(74..77) },
+    Punct { ch: ';', spacing: Alone, span: #0 bytes(77..78) }
+]
+    Finished dev [unoptimized + debuginfo] target(s) in 0.34s
+     Running `target/debug/foo`
+the answer was: 2
+```
+
+Here we can see how the input to the macro was a `TokenStream` representing
+the three input tokens `struct Foo;`. While our output only contained the
+`answer` function, we were still able to use `Foo` in the main program because
+custom derives *append* items, they don't replace them.
+
+Now this is a pretty wonky custom derive, and would likely be confusing to
+users! Custom derives are primarily geared towards implementing traits, like
+`Serialize` and `Deserialize`. The `syn` crate also has a [number of
+examples][synex] of defining custom derives.
+
+[procedural macros]: ../book/first-edition/procedural-macros.html
+[synex]: https://github.com/dtolnay/syn/tree/master/examples
+
+#### Custom attributes on Custom derive
+
+An additional feature of custom derive macros is that they can whitelist names
+of attributes which are considered not part of normal attribute macro expansion.
+Taking our example from earlier we can define:
+
+```rust,ignore
+#[proc_macro_derive(MyTrait, attributes(my_attribute))]
+pub fn foo(item: TokenStream) -> TokenStream {
+    // ...
+}
+```
+
+The extra `attributes` key in the `proc_macro_derive` attribute contains a
+comma-separated list of identifiers. Each identifier is a whitelist of attributes
+that can be attached to items which also have `#[derive(MyTrait)]`. The
+attributes will not do anything but will be passed through to the `foo`
+procedural macro defined above.
+
+If we change our invocation to look like:
+
+```rust,ignore
+#[derive(MyTrait)]
+#[my_attribute(hello)]
+struct Foo;
+```
+
+you'll see that the `#[my_attribute(hello)]` attribute is fed through to the
+macro for processing.
+
+Custom attributes are often used to customize the behavior of custom derives,
+such as the `#[serde]` attribute for the `serde` crate.
+
+### Custom attributes
+
+The third and final form of procedural macros is custom attributes. Custom
+attributes allow you to define a custom `#[attr]`-style attribute which can be
+attached to items and generate wrappers and such. These macros are defined like:
+
+```rust,ignore
+#[proc_macro_attribute]
+pub fn foo(attr: TokenStream, item: TokenStream) -> TokenStream {
```

Comment:
The second argument is not necessarily an item if this attribute is on a statement or expression. I would name the two `args` and `input`. Their role is fully analogous to command-line arguments and command-line input such as in:

```bash
$ input | ./foo a b c
```

```rust
#[foo(a, b, c)]
struct Input;
```

where `a b c` are arguments and separately there is input (the "in" in "stdin").